<launch>
    <!-- Our Goal-to-Velocity PID Controller -->
    <!-- <node pkg="uav_autonomous_utils" type="goal_to_velocity_pid.py" name="goal_to_velocity_pid" output="screen">
        <rosparam param="kp">[0.6, 0.6, 0.6, 0.8]</rosparam>
        <rosparam param="ki">[0.08, 0.08, 0.08, 0.0]</rosparam>
        <rosparam param="kd">[0.1, 0.1, 0.1, 0.1]</rosparam>
        <param name="max_speed" value="1.5" />
        <param name="max_yaw_rate" value="2.5" />
        <param name="goal_tolerance" value="0.1" />
        <param name="enable_yaw_alignment" value="true" />
        <param name="min_goal_change" value="0.2" />
        <remap from="/move_base_simple/goal" to="/move_base_visual"/>
    </node> -->

    <node pkg="uav_autonomous_utils" type="slidingModeController.py" name="pid_visual" output="screen">
        <rosparam param="kp">[0.5, 0.5, 0.5, 0.8]</rosparam>
        <rosparam param="ki">[0.06, 0.1, 0.1, 0.0]</rosparam>
        <rosparam param="kd">[0.1, 0.1, 0.1, 0.1]</rosparam>
        <param name="max_speed" value="2.5" />
        <param name="max_yaw_rate" value="2.0" />
        <param name="yaw_switch_distance" value="1.0" />
        <param name="position_tolerance" value="0.3" />
        <param name="yaw_tolerance" value="0.1" />
        <param name="is_gps_navigation_mode" value="false" />
        <remap from="/move_base_simple/goal" to="/move_base_visual"/>
        <remap from="/uav/odometry" to="/mavros/local_position/odom"/>
        <remap from="/uav/attractive_velocity" to="/uav/attractive_velocity"/>
    </node>

    <!-- Artificial Potential Fields Local Planner -->
    <param name="potential_fields/UAV_radius" value="0.1" />
    <node name="apf_node_visual" pkg="artificial_potential_fields" type="apf" output="screen">
        <param name="k_repulsive" value="0.0025" />
        <remap from="/uav/odometry" to="/mavros/local_position/odom"/>
        <remap from="/uav/attractive_velocity" to="/uav/attractive_velocity"/>
        <remap from="/laser/scan" to="/scan_from_depth_camera"/>
        <remap from="/potential_fields/velocity" to="/mavros/setpoint_velocity/cmd_vel"/>
    </node>

    <!-- Depth Image to Laser Scan Converter (dependency for APF) -->
    <node pkg="depthimage_to_laserscan" type="depthimage_to_laserscan" name="depthimage_to_laserscan_node_visual">
        <remap from="image" to="/iris/camera/depth/image_raw"/>
        <remap from="camera_info" to="/iris/camera/depth/camera_info"/>
        <remap from="scan" to="/scan_from_depth_camera"/>
        <param name="output_frame_id" value="iris/camera_link"/>
    </node>

        <!-- APF LOCAL PLANNER (LINEAR VELOCITY CALCULATOR) - REPLACED BY VFH3D -->
    <!-- 
    <node pkg="apf_local_planner" type="apf_local_planner" name="apf_local_planner_node" output="screen">
        <remap from="/lidar/livox" to="/iris/camera/depth/points"/>
        <remap from="/planner/goal" to="/move_base_visual"/>
        <param name="manager/max_vel" value="1.0"/>
        <param name="grid_map/obstacles_inflation" value="0.05" type="double"/>
        <param name="apf/sensor_max_range" value="5.0" type="double"/>
        <param name="apf/k_push" value="1.0" type="double"/>
        <param name="apf/k_push_xy" value="17.0" type="double"/>
        <param name="apf/k_push_z" value="0.0" type="double"/>
        <param name="apf/k_att" value="1.0" type="double"/>
        <param name="apf/min_dist" value="0.45" type="double"/>
        <param name="apf/max_att_dist" value="3.0" type="double"/>
        <param name="grid_map/ground_height" value="0.1" type="double"/>
        <param name="apf/ground_safe_height" value="0.3" type="double"/>
        <param name="apf/safe_distance" value="0.3" type="double"/>
    </node>
    -->

</launch>